libs/cuda_support/include/hpx/cuda_support/cublas_executor.hpp

```diff
@@ -0,0 +1,252 @@
+        {
+            switch (error)
+            {
+            case CUBLAS_STATUS_SUCCESS:
+                return "CUBLAS_STATUS_SUCCESS";
+            case CUBLAS_STATUS_NOT_INITIALIZED:
+                return "CUBLAS_STATUS_NOT_INITIALIZED";
+            case CUBLAS_STATUS_ALLOC_FAILED:
+                return "CUBLAS_STATUS_ALLOC_FAILED";
+            case CUBLAS_STATUS_INVALID_VALUE:
+                return "CUBLAS_STATUS_INVALID_VALUE";
+            case CUBLAS_STATUS_ARCH_MISMATCH:
+                return "CUBLAS_STATUS_ARCH_MISMATCH";
+            case CUBLAS_STATUS_MAPPING_ERROR:
+                return "CUBLAS_STATUS_MAPPING_ERROR";
+            case CUBLAS_STATUS_EXECUTION_FAILED:
+                return "CUBLAS_STATUS_EXECUTION_FAILED";
+            case CUBLAS_STATUS_INTERNAL_ERROR:
+                return "CUBLAS_STATUS_INTERNAL_ERROR";
+            case CUBLAS_STATUS_NOT_SUPPORTED:
+                return "CUBLAS_STATUS_NOT_SUPPORTED";
+            case CUBLAS_STATUS_LICENSE_ERROR:
+                return "CUBLAS_STATUS_LICENSE_ERROR";
+            }
+            return "<unknown>";
+        }
+    }    // namespace detail
+
+    // -------------------------------------------------------------------------
+    // exception type for failed launch of cuda functions
+    struct HPX_EXPORT cublas_exception : hpx::exception
+    {
+        cublas_exception(const std::string& msg, cublasStatus_t err)
+          : hpx::exception(hpx::bad_function_call, msg)
+          , err_(err)
+        {
+        }
+        cublasStatus_t get_cublas_errorcode()
+        {
+            return err_;
+        }
+
+    protected:
+        cublasStatus_t err_;
+    };
+
+    inline cublasStatus_t check_cublas_error(cublasStatus_t err)
+    {
+        if (err != CUBLAS_STATUS_SUCCESS)
+        {
+            std::stringstream temp;
+            temp << "cublas function returned error code :"
+                 << detail::_cublasGetErrorString(err);
+            throw cublas_exception(temp.str(), err);
+        }
+        return err;
+    }
+
+    namespace detail {
+        // specialization for return type of cublasStatus_t
+        template <typename... Args>
+        struct dispatch_helper<cublasStatus_t, Args...>
+        {
+            inline cublasStatus_t operator()(
+                cublasStatus_t (*f)(Args...), Args... args)
+            {
+                cublasStatus_t err = f(args...);
+                return check_cublas_error(err);
+            }
+        };
+
+    }    // namespace detail
+
+    // -------------------------------------------------------------------------
+    // a simple cublas wrapper helper object that can be used to synchronize
+    // cublas calls with an hpx future.
+    // -------------------------------------------------------------------------
+    struct cublas_executor : cuda_executor
+    {
+        // construct a cublas stream
+        cublas_executor(std::size_t device, bool event_mode = false)
```

Comment:
A parameter for the pointer mode should also be passed to the executor. https://docs.nvidia.com/cuda/cublas/index.html#scalar-parameters 

The default for most CUBLAS calls is `CUBLAS_POINTER_MODE_HOST`, if the CUBLAS function has an OUT parameter, the mode will actually synchronize with the stream on the thread the task is running, circumventing the `cudaEventQuery()` mechanism and blocking the thread on which the task is running. For such routines, it is good to have the option to change the mode to `CUBLAS_POINTER_MODE_DEVICE`.

Reply:
Thanks. I will add the pointer mode flag to the executor.

Reply:
I would suggest not to use a `bool` but the explicit CUDA constants. I also think there should be a default for that parameter (whatever is more commonly used).

Reply:
oops. I had a default in there originally, but I removed it when debugging and forgot to put it back. The bool param was an enum originally, but since there were only two, I made them `true_type` and `false_type` - would you prefer an enum?

Reply:
I misunderstood the comments. I had tag type for the event_mode/callback_mode - then switched it explicity to true_type/false_type - not for the pointer mode - please ignore my comment above.

The pointer mode in DLAF is being set on every cublas call - do you think that is actually necessary? Are you not in fact using the same mode for the whole application and could just set it once? do you think it is likely that you'll have it device mode in one executor and host mode in another?

Reply:
Since your handle is created and destroyed within the executor it can be set once when the executor is created. The mode depends on the function you are calling, a function like cublas<t>dot() with CUBLAS_POINTER_MODE_HOST will do a `cudaStreamSynchronize()` because with that pointer mode, the OUT parameter (`float *result`) is expected to be available on the CPU when the call completes. With CUBLAS_POINTER_MODE_DEVICE, the OUT parameter is instead on the GPU and no synchronization happens. This section of the docs discusses it : https://docs.nvidia.com/cuda/cublas/index.html#scalar-parameters

> do you think it is likely that you'll have it device mode in one executor and host mode in another?

Yes because I might be schedule a `cublas<t>gemm()` with host mode on one (there are no scalar OUT parameters) and `cublas<t>dot()` with device mode (where there is a scalar OUT parameter, i.e. the final value of the dot product).